Kubernetes Auth Method documentation not compatible with Kubernetes v1.21+

## Describe the bug

Kubernetes v1.21 introduced new bound service account tokens with following new properties:

1. Tokens are bound to a pod and they get revoked when pod is deleted.
2. Tokens will expire.
3. Tokens are audience-limited.

The Kubernetes Auth Method documentation does not cover the use of the new tokens. This results in failed authentication.

## To Reproduce

Pre-condition:

Deploy Vault inside Kubernetes. Configure Kubernetes Auth Method according to examples at https://www.vaultproject.io/docs/auth/kubernetes. Set JWT value to `token_reviewer_jwt="$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)"`.

**(1) Authentication fails because of invalidated token**

1. Delete the pod that runs Vault and wait for it to be recreated.
2. Try to authenticate using Kubernetes Auth Method
3. Vault returns `403 Forbidden` with `{ "errors": [ "permission denied" ] }` in response body.

Vault prints following error:

```
2021-10-18T12:09:01.075Z [ERROR] auth.kubernetes.auth_kubernetes_726dbb64: login unauthorized due to: lookup failed: service account unauthorized; this could mean it has been deleted or recreated with a new token
```

Kubernetes API server prints following error

```
E1018 12:09:01.074341       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, Token has been invalidated]"
```

**(2) Authentication fails because of expired token**

1. Configure Kubernetes cluster with a short expiration time by setting API server parameter `service-account-max-token-expiration` to `1h`. Wait for 1 hour for the token to expire .
2. Try to authenticate using Kubernetes Auth Method. 
3. Vault returns `403 Forbidden` with `{ "errors": [ "permission denied" ] }` in response body.

Vault prints following error:

```
2021-10-18T11:47:23.756Z [ERROR] auth.kubernetes.auth_kubernetes_97f02f17: login unauthorized due to: lookup failed: service account unauthorized; this could mean it has been deleted or recreated with a new token
```

Kubernetes API server prints following error

```
E1018 11:47:23.755730       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, Token has expired.]"
```


## Proposed fix

Include following PR
* https://github.com/hashicorp/vault-plugin-auth-kubernetes/pull/122

Update documentation, taking following points into account:

* It is recommended NOT to store Kubernetes service account token persistently into config when running Vault inside Kubernetes. Instead, the recommendation is to use the local token and CA files. 
* When running Vault outside of Kubernetes, user needs to use legacy token which is not bound to any pod and does not expire. User should be instructed how to access legacy token via Secret directly. User should be WARNED that the legacy tokens could be removed in later Kubernetes release completely.

I also think following point would be beneficial

* Recommend `disable_iss_validation=True`. The benefit is that user does not need to find the correct `iss` from their specific cluster (or change it if it changes later). Kubernetes Token Review API will validate the token, so I do not see negative side in recommending that.

--- a/changelog/13595.txt
+++ b/changelog/13595.txt
@@ -1,0 +1,2 @@
+```release-note:improvement
+auth/kubernetes: Added support for short-lived tokens for Kubernetes 1.21+ compatibility

--- a/changelog/13595.txt
+++ b/changelog/13595.txt
@@ -1,2 +1,3 @@
 ```release-note:improvement
-auth/kubernetes: Added support for short-lived tokens for Kubernetes 1.21+ compatibility
+auth/kubernetes: Added support for dynamically reloading short-lived tokens for better Kubernetes 1.21+ compatibility
+```

--- a/go.mod
+++ b/go.mod
@@ -96,7 +96,7 @@ require (
 	github.com/hashicorp/vault-plugin-auth-gcp v0.11.3
 	github.com/hashicorp/vault-plugin-auth-jwt v0.11.4
 	github.com/hashicorp/vault-plugin-auth-kerberos v0.5.0
-	github.com/hashicorp/vault-plugin-auth-kubernetes v0.11.3
+	github.com/hashicorp/vault-plugin-auth-kubernetes v0.7.1-0.20220107030939-d289258274b7
 	github.com/hashicorp/vault-plugin-auth-oci v0.9.0
 	github.com/hashicorp/vault-plugin-database-couchbase v0.5.1
 	github.com/hashicorp/vault-plugin-database-elasticsearch v0.9.1

--- a/website/content/api-docs/auth/kubernetes.mdx
+++ b/website/content/api-docs/auth/kubernetes.mdx
@@ -29,8 +29,10 @@ access the Kubernetes API.
 
 - `kubernetes_host` `(string: <required>)` - Host must be a host string, a host:port pair, or a URL to the base of the Kubernetes API server.
 - `kubernetes_ca_cert` `(string: "")` - PEM encoded CA cert for use by the TLS client used to talk with the Kubernetes API. NOTE: Every line must end with a newline: `\n`
+  If not set, the local CA cert will be used if running in a Kubernetes pod.
 - `token_reviewer_jwt` `(string: "")` - A service account JWT used to access the TokenReview
   API to validate other JWTs during login. If not set,
+  the local service account token is used if running in a Kubernetes pod, otherwise
   the JWT submitted in the login payload will be used to access the Kubernetes TokenReview API.
 - `pem_keys` `(array: [])` - Optional list of PEM-formatted public keys or certificates
   used to verify the signatures of Kubernetes service account

--- a/website/content/docs/auth/kubernetes.mdx
+++ b/website/content/docs/auth/kubernetes.mdx
@@ -125,9 +125,8 @@ Kubernetes auth:
 * The value of the JWT's `"iss"` claim depends on the cluster's configuration.
 
 The changes to token lifetime are important when configuring the
-[`token_reviewer_jwt`](/api-docs/auth/kubernetes#token_reviewer_jwt) option. You
-must avoid using a short-lived token because Vault stores that token in Vault
-storage and does not automatically refresh it. If a short-lived token is used,
+[`token_reviewer_jwt`](/api-docs/auth/kubernetes#token_reviewer_jwt) option.
+If a short-lived token is used,
 Kubernetes will revoke it as soon as the pod or service account are deleted, or
 if the expiry time passes, and Vault will no longer be able to use the
 `TokenReview` API. See [How to work with short-lived Kubernetes tokens][short-lived-tokens]
@@ -155,6 +154,7 @@ table summarizes the options, each of which is explained in more detail below.
 
 | Option                               | All tokens are short-lived | Can revoke tokens early | Other considerations |
 | ------------------------------------ | -------------------------- | ----------------------- | -------------------- |
+| Use local token as reviewer JWT      | Yes                        | Yes                     | Requires Vault VERSION_TODO |
 | Use client JWT as reviewer JWT       | Yes                        | Yes                     | Operational overhead |
 | Use long-lived token as reviewer JWT | No                         | Yes                     |                      |
 | Use JWT auth instead                 | Yes                        | No                      |                      |
@@ -167,6 +167,23 @@ short-lived tokens. If you would like to disable this, set
 [here](/docs/auth/jwt/oidc_providers#specifying-ttl-and-audience) for an example.
 
 [k8s-extended-tokens]: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/#options
+
+#### Use local service account token as the reviewer JWT
+
+When running Vault in a Kubernetes pod the recommended option is to use local
+service account token. Vault will periodically re-read the file to support
+short-lived tokens. Local token and CA certificate are used by configuring the
+auth method without setting `token_reviewer_jwt` and `kubernetes_ca_cert`.
+
+```bash
+vault write auth/kubernetes/config \
+    kubernetes_host=https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_PORT
+```
+
+!> **Note:** Requires Vault VERSION_TODO. In earlier versions the service account
+token and CA certificate is read once and stored in Vault storage.
+When the service account token expires or is revoked, Vault will no longer be
+able to use the `TokenReview` API and client authentication will fail.
 
 #### Use the Vault client's JWT as the reviewer JWT
 

--- a/website/content/docs/auth/kubernetes.mdx
+++ b/website/content/docs/auth/kubernetes.mdx
@@ -154,7 +154,7 @@ table summarizes the options, each of which is explained in more detail below.
 
 | Option                               | All tokens are short-lived | Can revoke tokens early | Other considerations |
 | ------------------------------------ | -------------------------- | ----------------------- | -------------------- |
-| Use local token as reviewer JWT      | Yes                        | Yes                     | Requires Vault VERSION_TODO |
+| Use local token as reviewer JWT      | Yes                        | Yes                     | Requires Vault (1.10+) to be deployed on the Kubernetes cluster |
 | Use client JWT as reviewer JWT       | Yes                        | Yes                     | Operational overhead |
 | Use long-lived token as reviewer JWT | No                         | Yes                     |                      |
 | Use JWT auth instead                 | Yes                        | No                      |                      |
@@ -170,17 +170,19 @@ short-lived tokens. If you would like to disable this, set
 
 #### Use local service account token as the reviewer JWT
 
-When running Vault in a Kubernetes pod the recommended option is to use local
+When running Vault in a Kubernetes pod the recommended option is to use the pod's local
 service account token. Vault will periodically re-read the file to support
-short-lived tokens. Local token and CA certificate are used by configuring the
-auth method without setting `token_reviewer_jwt` and `kubernetes_ca_cert`.
+short-lived tokens. To use the local token and CA certificate, omit
+`token_reviewer_jwt` and `kubernetes_ca_cert` when configuring the auth method.
+Vault will attempt to load them from `token` and `ca.crt` respectively inside
+the default mount folder `/var/run/secrets/kubernetes.io/serviceaccount/`.
 
 ```bash
 vault write auth/kubernetes/config \
     kubernetes_host=https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_PORT
 ```
 
-!> **Note:** Requires Vault VERSION_TODO. In earlier versions the service account
+!> **Note:** Requires Vault 1.10+. In earlier versions the service account
 token and CA certificate is read once and stored in Vault storage.
 When the service account token expires or is revoked, Vault will no longer be
 able to use the `TokenReview` API and client authentication will fail.

